installers/Config.hs

```diff
@@ -0,0 +1,134 @@
+
+diagReadCaseInsensitive :: (Bounded a, Enum a, Read a, Show a) => String -> Maybe a
+diagReadCaseInsensitive str = diagRead $ Data.Char.toLower <$> str
+  where mapping    = Map.fromList [ (Data.Char.toLower <$> show x, x) | x <- enumFromTo minBound maxBound ]
+        diagRead x = Just $ flip fromMaybe (Map.lookup x mapping)
+                     (error $ printf ("Couldn't parse '%s' as one of: %s")
+                                     str (L.intercalate ", " $ Map.keys mapping))
+
+optReadLower :: (Bounded a, Enum a, Read a, Show a) => ArgName -> ShortName -> Optional HelpMessage -> Parser a
+optReadLower = opt (diagReadCaseInsensitive . unpack)
+argReadLower :: (Bounded a, Enum a, Read a, Show a) => ArgName -> Optional HelpMessage -> Parser a
+argReadLower = arg (diagReadCaseInsensitive . unpack)
+
+data Options = Options
+  { oAPI           :: API
+  , oBuildJob      :: Maybe BuildJob
+  , oCertPass      :: Maybe Text
+  , oCluster       :: Cluster
+  , oDaedalusVer   :: Version
+  , oOutput        :: Text
+  , oPullReq       :: Maybe PullReq
+  , oTestInstaller :: TestInstaller
+  , oCI            :: CI
+  }
+
+optionsParser :: Parser Options
+optionsParser = Options
+  <$> (fromMaybe Cardano <$> (optional $
+                optReadLower "api"                 'a' "Backend API:  cardano or etc"))
+  <*> (optional   $
+      (BuildJob  <$> optText "build-job"           'b' "CI Build Job/ID"))
+  <*> (optional   $
+                     optText "cert-pass"           'p' "Certificate password")
+  <*> (fromMaybe Mainnet <$> (optional $
+                optReadLower "cluster"             'c' "Cluster the resulting installer will target:  mainnet or staging"))
+  <*> (fromMaybe "dev"   <$> (optional $
+      (Version   <$> optText "daedalus-version"    'v' "Daedalus version string")))
+  <*>                optText "output"              'o' "Installer output file"
+  <*> (optional   $
+      (PullReq   <$> optText "pull-request"        'r' "Pull request #"))
+  <*> flag TestInstaller     "test-installer"      't' "Test installers after building"
+  <*> pure Buildkite -- NOTE: this is filled in by auto-detection
+
+
+
+lshow :: Show a => a -> Text
+lshow = pack . fmap Data.Char.toLower . show
+
+dhallTopExpr :: Text -> Config -> OS -> Cluster -> Text
+dhallTopExpr path Launcher os cluster = path <> "/launcher.dhall ( "<>path<>"/" <> lshow cluster <> ".dhall "<>path<>"/" <> lshow os <> ".dhall ) "<>path<>"/" <> lshow os <> ".dhall"
+dhallTopExpr path Topology os cluster = path <> "/topology.dhall ( "<>path<>"/" <> lshow cluster <> ".dhall "<>path<>"/" <> lshow os <> ".dhall )"
+
+generateConfig :: Request -> FilePath -> FilePath -> IO ()
+generateConfig Request{..} configRoot outFile = handle $ do
+  GHC.IO.Encoding.setLocaleEncoding GHC.IO.Encoding.utf8
+
+  Dhall.detailed $ do
+    let inText = LT.fromStrict $ dhallTopExpr (pack configRoot) rConfig rOS rCluster
+    print inText
```

Comment:
This shouldn't be here right?

Reply:
@domenkozar, I found it helpful to see the input expression printed to stdout, to help ensuring things are as they are at key junctions -- since we're dealing with `sh` and `cmd.exe` here..

As an example: https://buildkite.com/input-output-hk/daedalus/builds/429#719c10ca-f6f5-4a4c-8300-492977753b64/22-509